### ISSUE TITLE
MPC not fully reliable

### ISSUE THREAD

**Issue by j-brendel** [rl-institut/smooth#212]
Without MPC: Battery from solar or from grid only:

![Opt_1D1h_sum_maxPV](https://user-images.githubusercontent.com/39203376/99984877-e638b400-2dad-11eb-9a0b-b123f951d604.png)

![Opt_1D1h_sum_minBat](https://user-images.githubusercontent.com/39203376/99984892-e9cc3b00-2dad-11eb-809f-f192e120ef03.png)

Preview of possibility with MPC:
![Opt_1D1h_sum_MPCtest](https://user-images.githubusercontent.com/39203376/99984746-c30e0480-2dad-11eb-844d-c1109279f6d9.png)

**Comment by j-brendel** [comment]
To try out:
[UHAL_v2_MPC.zip](https://github.com/rl-institut/smooth/files/5584205/UHAL_v2_MPC.zip)

Use julian_working_branch of smooth

**Comment by j-ti** [comment]
Grund, warum das Modell oben bei langem Zeithorizont mit linearem MPC nicht lösbar ist, liegt nach meiner Analyse genau an der Linearität die bei nahezuvollem Akku nicht angenommen werden kann. Genauer: Wenn E_max (charging) beim ersten Zeitschritt sehr sehr gering ist, weil die Batterie gerade voll ist, gilt der Wert in einer linearer MPC  für den gesamten Zeithorizont auch wenn zwischendurch wieder entladen wird.

Um auf den Vorteil von Linearität nicht verzichten zu müssen und es schnell umzusetzen, schlage ich vor:
1. die Batterie nicht maximal ausreizen (SOC_min, SOC_max) evtl. sowieso besser für die Lebensdauer
2. ein desired SOC einführen mit entsprechenden `artificial_costs`, so dass generell nicht bis SOC_max geladen wird, aber  weiterhin die Möglichkeit besteht, wenn es unbedingt nötig ist (auch besser für die Lebensdauer)

Ob das für dich so ausreicht müssen wir nochmal am Donnerstag besprechen

**Comment by j-ti** [comment]
![MPC_hz20_desiredSOC](https://user-images.githubusercontent.com/56872869/100760127-b747c100-33f1-11eb-83ed-314456f56f71.png)

**Comment by j-ti** [comment]
'battery_capacity_bt2': 300 * 1e3
'mpc_control_horizon': 30,

within battery component:
`self.p_in_max = (self.c_rate_charge * self.battery_capacity
            ) / self.efficiency_charge`
`self.p_out_max = (self.c_rate_discharge * self.battery_capacity)\
                         * self.efficiency_discharge`

![MPC_hz30_Pmax_nolosscompencation](https://user-images.githubusercontent.com/56872869/101780856-5f970d00-3af7-11eb-95d5-ea206de44ddd.png)

**Comment by j-ti** [comment]
same as above but:
'battery_capacity_bt2': 230 * 1e3

![MPC_hz30_Pmax_nolosscompencation_Cap230](https://user-images.githubusercontent.com/56872869/101781122-b997d280-3af7-11eb-9a32-590a1bbf53e4.png)

**Comment by j-ti** [comment]
here again 'battery_capacity_bt2': 300 * 1e3
with the P max adaptation, the desired SOC with low_vac are not necessary:
switch default vac back to "keep Battery full (from grid) to minimize Bat":
`    'vac_in': -0.20 / 1000,
    'vac_out': 0.20 / 1000,`

![MPC_hz30_Pmax_nolosscompencation_noDesiredSOC_optBatCap](https://user-images.githubusercontent.com/56872869/101781268-de8c4580-3af7-11eb-8cad-68c9a7de8b29.png)

**Comment by j-ti** [comment]
example:
162 | li_battery |   | capacity | 56 |   | 50000 | **218721,02** | 250000 | 
163 | li_battery |   | capacity | 57 |   | 50000 | 50000        | 250000 |  

(SOC_i*(1-loss_rate)- SOC_min) * eff_decharge = flow(bat->bel)
(218721,02*(1-0,067/100/24)- 50000) * 0,95 = 160279,168336282

In reality the value output from oemof is 160279,17
As for oemof it is the end result it is seems no problem in which direction it rounds.

In this example which uses the battery down to SOC_min, one can see if smooth takes this rounded value the energy will not suffice to compensate for grid limitations by only smaller than 0.005. Same could appear in the other direction with SOC_max

**Comment by j-ti** [comment]
As a test it is possible to add this value 0.005 Ah, which is very small in comparison to in this case E_min = 50 000 Ah
`self.soc = (df_storage[i_result][0]+0.005) / self.battery_capacity`
... then MPC works as expected:
I chose low feed_in tariff of 0.05, medium vac +/- 0.10 and the from grid cost of 0.181 per kW:

![mpc_horz40_bat250_roundingerror_resolved_lowfeedin](https://user-images.githubusercontent.com/56872869/102015779-ddf4e880-3d5d-11eb-906d-32908de3521b.png)

**Comment by j-ti** [comment]
Thie intermediate results for the mpc for the battery are as follows:
![mpc_horz40_bat250_roundingerror_resolved_lowfeedin_intermediateBat](https://user-images.githubusercontent.com/56872869/102015801-febd3e00-3d5d-11eb-90de-e63ca12d81d1.png)

**Comment by j-ti** [comment]
Obviously, to add this value is a hack. Although as 0.005 << 50000 (example E_bat,min) and the price per timestep is very low (0.181 / 1000) * 0.005 = 9.05 x 10^-7
and in case timestep is 1 hour it becomes an error per Year of
(0.181 / 1000) * 0.005 * 24 * 365 = 0.0079278
There should be searched for a better way. For a timestep of 1s the price per year rises to 2.854008

**Comment by j-ti** [comment]
As argued before the limitations for the battery's `self.p_in_max `and `self.p_out_max` as they were do not make sense as oemof already handles the constraints itself according to loss_rate, efficiency etc.

**Comment by j-ti** [comment]
Full charging in one timestep is possible for c_rate=1 as p_max=battery_capacity is always bigger than battery_capacity - battery_capacity* SOC_min

**Comment by j-ti** [comment]
> As a test it is possible to add this value 0.005 Ah, which is very small in comparison to in this case E_min = 50 000 Ah
> `self.soc = (df_storage[i_result][0]+0.005) / self.battery_capacity`
> ... then MPC works as expected:
> I chose low feed_in tariff of 0.05, medium vac +/- 0.10 and the from grid cost of 0.181 per kW:
> 
> ![mpc_horz40_bat250_roundingerror_resolved_lowfeedin](https://user-images.githubusercontent.com/56872869/102015779-ddf4e880-3d5d-11eb-906d-32908de3521b.png)

For similar reasons as described in [issue #212 comment](https://github.com/rl-institut/smooth/issues/218#issuecomment-744017840 ), the vac_in/out and feed in tariff should be chosen so that the negative cost of charing and discharging is not profitable when using solar energy, which has no variable cost.

I chose low feed_in tariff of -0.05 - 0.13 art_cost , medium vac +/- 0.10 and the from grid cost of 0.181 per kW:
This way charging and discharging does not appear between timesteps 125 - 140:

![mpc_horz40_bat250_roundingerror_resolved](https://user-images.githubusercontent.com/56872869/102091708-7f903e80-3e1f-11eb-81be-e435b2f70ecb.png)
![mpc_linear_intermediate_plots](https://user-images.githubusercontent.com/56872869/102091722-83bc5c00-3e1f-11eb-88cc-290031c06667.png)
